### ISSUE TITLE
How to set the learning rate in CNTK?

### ISSUE THREAD

**Issue by NewEricWang** [Microsoft/CNTK#322]
Hi, 
       I am trying to use the CNTK to train a FNN model. I find the change of the learning rate is abnormal. Some log is as follows:
       Finished Epoch[ 1 of 30]: [Training Set] TrainLossPerSample = 5.5450029; TotalSamplesSeen = 24402857; EvalErrPerSample = 0.78378993; AvgLearningRatePerSample = 0.0046875002; EpochTime=4344.08
Finished Epoch[ 2 of 30]: [Training Set] TrainLossPerSample = 5.5427551; TotalSamplesSeen = 48805714; EvalErrPerSample = 0.78377134; AvgLearningRatePerSample = 0.0046875002; EpochTime=4351.05
Finished Epoch[ 3 of 30]: [Training Set] TrainLossPerSample = 5.5427213; TotalSamplesSeen = 73208571; EvalErrPerSample = 0.78377372; AvgLearningRatePerSample = 0.0046875002; EpochTime=4358.21
Finished Epoch[ 4 of 30]: [Training Set] TrainLossPerSample = 5.5428209; TotalSamplesSeen = 97611428; EvalErrPerSample = 0.78377092; AvgLearningRatePerSample = 0.0046875002; EpochTime=4360.92
Finished Epoch[ 5 of 30]: [Training Set] TrainLossPerSample = 5.5427647; TotalSamplesSeen = 122014285; EvalErrPerSample = 0.783764; AvgLearningRatePerSample = 0.0046875002; EpochTime=4360.92
Finished Epoch[ 6 of 30]: [Training Set] TrainLossPerSample = 5.5391655; TotalSamplesSeen = 146417142; EvalErrPerSample = 0.78374195; AvgLearningRatePerSample = 0.0015625; EpochTime=3356.86
Finished Epoch[ 7 of 30]: [Training Set] TrainLossPerSample = 5.5391321; TotalSamplesSeen = 170819999; EvalErrPerSample = 0.78374302; AvgLearningRatePerSample = 0.0015625; EpochTime=3361.55
Finished Epoch[ 8 of 30]: [Training Set] TrainLossPerSample = 5.5383992; TotalSamplesSeen = 195222856; EvalErrPerSample = 0.78374302; AvgLearningRatePerSample = 0.00093750001; EpochTime=3360.79
Finished Epoch[ 9 of 30]: [Training Set] TrainLossPerSample = 5.5317965; TotalSamplesSeen = 219625713; EvalErrPerSample = 0.78380466; AvgLearningRatePerSample = 0.00056250001; EpochTime=3356.51
Finished Epoch[10 of 30]: [Training Set] TrainLossPerSample = 4.7810326; TotalSamplesSeen = 244028570; EvalErrPerSample = 0.6621576; AvgLearningRatePerSample = 0.00033750001; EpochTime=3360.03

I don't understand why the learning rate change from 0.0046875002 into 0.0015625 in epoch 6. It doesn't meet my expectations. My configure about learning rate is as follows:
    SGD = [ 
        epochSize=0
        # epochSize = 20480
        minibatchSize = 256*5:512*7:1024
        learningRatesPerMB = 1.2*5:0.8
        # autoLearnRateSearchType = true
        numMBsToShowResult = 10
        momentumPerMB = 0.9*10:0.8*10:0.7
        maxEpochs = 30
        dropoutRate=0.0    
        keepCheckPointFiles = false
    
        # new add by hlwang
        frameMode = true    
        # Additional optional parameters are: parallelizationStartEpoch
        parallelTrain = [ 
            parallelizationMethod = "DataParallelSGD"
            distributedMBReading = true
    
            # Additional optional parameters are: useZeroThresholdFor1BitQuantization
            dataParallelSGD = [ 
                gradientBits = 1 
            ]
        ]
        autoAdjust=[
            autoAdjustLR="adjustAfterEpoch"
            reduceLearnRateIfImproveLessThan=0.5
            loadBestModel=true
            increaseLearnRateIfImproveMoreThan=1000000000
            learnRateDecreaseFactor=0.6
            learnRateIncreaseFactor=1.382
        ]
    ]

**Comment by yzhang87** [comment]
I think this is because you set this block:
 autoAdjust=[
        autoAdjustLR="adjustAfterEpoch"
        reduceLearnRateIfImproveLessThan=0.5
        loadBestModel=true
        increaseLearnRateIfImproveMoreThan=1000000000
        **learnRateDecreaseFactor=0.6**
        learnRateIncreaseFactor=1.382
    ]

It means if the training criterion become worse, the learning rate will decrease.

thanks,
Yu

**Comment by NewEricWang** [comment]
Hi, Yu,
      I get it. I see the learning rate 0.00093750001 in epoch 8 is equal with 0.0015625*0.6. But  the learning rate 0.0015625 in epoch 6 is not equal with 0.0046875002*0.6. Could you tell me the reason?
Thanks!

**Comment by frankseide** [comment]
It does not seem to be correct indeed. The factor for this one step is 1/3, not 0.6. Strangely, the runtime goes down at the same point, as if something changed in the config.

**Comment by NewEricWang** [comment]
Hi, frankseide
     I confirm the config isn't changed in training procedure. In the same training and cv data set, I get most worse results (i.e. EvalErrPerSample = 0.60615587 on training data and EvalErrPerSample = 0.62080282 on validation set) by using above SGD config compared to following SGD config:
         SGD = [ 
        epochSize=0
        minibatchSize = 256*5:512*7:1024
        learningRatesPerMB = 1.0*2:0.8*20:0.01
        numMBsToShowResult = 10
        momentumPerMB = 0.9*10:0.8*10:0.7
        maxEpochs = 30
        dropoutRate=0.0    
        keepCheckPointFiles = false
        frameMode = true
        parallelTrain = [ 
            parallelizationMethod = "DataParallelSGD"
            distributedMBReading = true
            dataParallelSGD = [ 
                gradientBits = 1 
            ]
        ]
        autoAdjust=[
            autoAdjustLR="adjustAfterEpoch"
            reduceLearnRateIfImproveLessThan=0.001
            loadBestModel=true
            increaseLearnRateIfImproveMoreThan=1000000000
            learnRateDecreaseFactor=0.6
            learnRateIncreaseFactor=1.382
        ]
    ]   

   Using the config, I can get EvalErrPerSample = 0.30201286 on training data and EvalErrPerSample = 0.37176935 on validation set. 
    How should I set the config about learning rate to get better results? Why is the learning rate  reduced by 1/3 factor in epoch 6?
